keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+non-empty `.metadata.ownerReferences`). You can think of the resources that
+don't have owners as root nodes in a DAG (directed acyclic graph). Those
+resources without owners may never be deleted after they're no longer needed. 
+
+Most of the Kubernetes resources that users create directly won't have owners. 
+
+#### Immutable Pattern: Short-lived, Frequently-created Resources
+
+Short-lived resources, such as Jobs (that are not managed by CronJobs), Secrets
+and ConfigMaps, are often treated as _immutable_ resources -- they are
+frequently created and not updated by users. We also have discussed making such
+resources immutable ([#10179][]).
+
+However, it's difficult for the users to clean up those resources automatically,
+because those resources don't have owners. Without a way to clean up those
+frequently created, short-lived resources, those resources can accumulate and
+overload a Kubernetes cluster very easily. Even if we can avoid the overload
+issue by implementing a cluster-wide (global) resource quota, users won't be
+able to create new resources without cleaning up old ones first. 
+
+[#10179]: https://github.com/kubernetes/kubernetes/issues/10179#issuecomment-384019058
+
+#### Cannot Centralize Cleanup of Multiple Resources with OwnerReferences
+
+Moreover, people building declarative application description with several
+Kubernetes resources cannot easily centralize clean up using owner references,
+meaning, to create a bunch of objects and deleting one of them to make the other
+objects to go away with it. 
+
+One of the reasons is that owner resource [UID][] is required when specifying
+owner references, but UID is unknown until after a resource is created. Even if
+Kubernetes removes this limitation and allow named based owner references,
+there’s still a resource creation ordering problem: any child resource created
+before its parent resource is prone to be cleaned up by the garbage collector
+unexpectedly.
+
+What's more, a child resource may depend on many other parent resources. It's
+likely for the list of owners of a child resource to be excessively long, and
+hit Kubernetes resource size limit. The resource then cannot be updated anymore
+and becomes useless. 
+
+[UID]: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids
+
+## Proposal
+
+The design has three main parts: 
+
+1. **Object-lifecycle-respecting TTL mechanism.** Provide observability before
+   deletion, and address the resource creation ordering problems. See [detailed
+   design of TTL][].
+1. **Garbage collect resources that finish execution.** Delete a resource object
+   when it finishes execution (either terminates or completes). See [detailed
+   design of GC after use (part 1)][].
+1. **Garbage collect resources that are not used by other resources.** Extract
+   usage references of a resource object and delete the object when it is no
+   longer referenced. Garbage collector will count usage references and address
```

Comment:
inverting the graph direction garbage collection is responsible for has some significant implications:
* latency or disruption in the GC controller's informers means that new usages of a `GCIfUnused: true` resource are unobserved, and can incorrectly treat that resource as unused and delete it
* any inability to discover or reach aggregated APIs has to halt GC of `GCIfUnused: true` resources until resolved, since the undiscovered resources could contain usages that would prevent deletion

Reply:
GC controller can do a sanity check before deleting a resource. This check could be expensive, though. 

It'd be useful for GCIfUnused resources to specify the api group versions of its potential users. This would make queries less expensive and make unreachable aggregated API only affect GCIfUnused resources that cares about the API.

Reply:
To solve this problem, I suggested making the GC keep a per-group-version used/not used marker on objects opting in to this mechanism.

If the apiserver serving the group-version is down, it's true that GC can't observe new uses, but also true that people can't create new uses. So this leaves only a small window where someone adds a new use and then the apiserver goes down before GC observes it. This is likely preferable to blocking the whole GC (which could render the control plane inoperable) if an apiserver is down.